pmlc/target/x86/pipeline.cc

```diff
@@ -104,25 +84,124 @@ struct ConvertStandardToLLVMPass
     target.addDynamicallyLegalOp<omp::ParallelOp>([&](omp::ParallelOp op) {
       return typeConverter.isLegal(&op.getRegion());
     });
     target.addLegalOp<omp::TerminatorOp, omp::TaskyieldOp, omp::FlushOp,
                       omp::BarrierOp, omp::TaskwaitOp>();
-    target.addDynamicallyLegalOp<LLVM::ExtractValueOp>(
-        [](LLVM::ExtractValueOp op) {
-          return !mlir::dyn_cast_or_null<LLVM::InsertValueOp>(
-              op.container().getDefiningOp());
-        });
     if (failed(applyPartialConversion(module, target, patterns))) {
       signalPassFailure();
     }
   }
 };
 
+// OpenMP has issues passing values through to OpenMP blocks.  As a workaround,
+// we have a simple pass to smuggle values that cross the boundary via an
+// alloca()'d struct.
+struct OpenMPWorkaroundPass final
+    : public OpenMPWorkaroundBase<OpenMPWorkaroundPass> {
+  void runOnOperation() final {
+    LLVM::LLVMFuncOp funcOp = getOperation();
+    OpBuilder builder{&getContext()};
+    funcOp.walk([&](omp::ParallelOp parOp) {
+      // The values used to construct the smuggling struct.
+      SmallVector<LLVM::LLVMType, 8> fieldTypes;
+      SmallVector<Value, 8> fieldSrcs;
+
+      // A mapping from original values to the OpOperands within the
+      // parallel region that use those values, for value replacement.
+      std::unordered_map<void *, std::list<OpOperand *>> valueMap;
+
+      // Find values that cross the loop
+      parOp.walk([&](mlir::Operation *childOp) {
+        if (childOp == parOp.getOperation()) {
+          return;
+        }
+        for (auto &opOperand : childOp->getOpOperands()) {
+          auto value = opOperand.get();
+
+          // Simple check: if it's not an LLVM type, or if it's an LLVM
+          // pointer type, we don't need or want to smuggle this value in
+          // via a struct.
+          auto llvmType = value.getType().dyn_cast<LLVM::LLVMType>();
+          if (!llvmType || llvmType.isPointerTy()) {
+            continue;
+          }
+
+          Region *region = nullptr;
+          if (auto blockArg = value.dyn_cast<BlockArgument>()) {
+            region = blockArg.getParentRegion();
+          } else {
+            region = value.getDefiningOp()->getParentRegion();
+          }
+
+          // Check the value's source's nested region chain: if we encounter the
+          // parOp's region, then the value is defined within the parOp and
+          // doesn't need to be (can't be) smuggled.
+          bool definedInsideParOp = false;
+          while (region) {
+            if (region == &parOp.region()) {
+              definedInsideParOp = true;
+              break;
+            }
+            region = region->getParentRegion();
+          }
+          if (definedInsideParOp) {
+            continue;
+          }
+
+          // Otherwise, we need to smuggle the value through an alloca'd struct.
+          auto [it, inserted] = valueMap.try_emplace(value.getAsOpaquePointer(),
```

Comment:
Please don't use C++17 features, I have to remove them when I go to integrate with downstream users.

Reply:
Which downstream users?  If this is going to be an ongoing thing, should we disable c++17 in our build?  It might be easier than trying to remember which features landed in which version...?